internal/distro/fedora/distro.go

```diff
@@ -196,49 +211,73 @@ func (t *imageType) PartitionType() string {
 	return ""
 }
 
-func (t *imageType) Packages(bp blueprint.Blueprint) ([]string, []string) {
-	packages := append(t.packages, bp.GetPackages()...)
-	timezone, _ := bp.Customizations.GetTimezoneSettings()
-	if timezone != nil {
-		packages = append(packages, "chrony")
+func (t *imageType) getPackages(name string) rpmmd.PackageSet {
+	getter := t.packageSets[name]
+	if getter == nil {
+		return rpmmd.PackageSet{}
+	}
+
+	return getter(t)
+}
+
+func (t *imageType) PackageSets(bp blueprint.Blueprint) map[string]rpmmd.PackageSet {
+	// merge package sets that appear in the image type with the package sets
+	// of the same name from the distro and arch
+	mergedSets := make(map[string]rpmmd.PackageSet)
+
+	imageSets := t.packageSets
+
+	for name := range imageSets {
+		mergedSets[name] = t.getPackages(name)
 	}
+
+	if _, hasPackages := imageSets[osPkgsKey]; !hasPackages {
+		// should this be possible??
+		mergedSets[osPkgsKey] = rpmmd.PackageSet{}
+	}
+
+	// every image type must define a 'build' package set
+	if _, hasBuild := imageSets[buildPkgsKey]; !hasBuild {
+		panic(fmt.Sprintf("'%s' image type has no '%s' package set defined", t.name, buildPkgsKey))
+	}
+
 	if t.bootable {
-		packages = append(packages, t.arch.bootloaderPackages...)
+		switch t.arch.Name() {
+		case distro.X86_64ArchName:
+			mergedSets[osPkgsKey] = mergedSets[osPkgsKey].Append(rpmmd.PackageSet{Include: t.arch.bootloaderPackages})
+		case distro.Aarch64ArchName:
+			mergedSets[osPkgsKey] = mergedSets[osPkgsKey].Append(rpmmd.PackageSet{Include: t.arch.bootloaderPackages})
+		}
+
```

Comment:
Since you introduced the `packageSetFunc()` and separated package sets into a `package_sets.go`, I think that it makes sense to go all the way and adopt the RHEL-8.6/9.0 approach completely. Which would mean e.g  substituting this section with https://github.com/osbuild/osbuild-composer/blob/b2d18166de636613f27b42c55aa2679461e9928b/internal/distro/rhel90/package_sets.go#L193-L238

IMO this will also make it easier to rework the Fedora distro definition to manifests V2...

Reply:
Understood.